src/lib/layout/index.spec.ts

```diff
@@ -135,33 +134,42 @@ describe('projectRoot', () => {
 
 describe('sourceRoot', () => {
   it('defaults to project dir', async () => {
-    ctx.setup({ 'tsconfig.json': '' })
-    const result = await ctx.createLayoutThrow()
-    expect(result.sourceRoot).toEqual('__DYNAMIC__')
-    expect(result.projectRoot).toEqual('__DYNAMIC__')
+    ctx.setup({ 'tsconfig.json': '', 'app.ts': '' })
+    const res = await ctx.createLayout().then(rightOrThrow)
+    expect(res.sourceRoot).toEqual('__DYNAMIC__')
+    expect(res.projectRoot).toEqual('__DYNAMIC__')
   })
-  it('honours the value in tsconfig rootDir', async () => {
-    ctx.setup({ 'tsconfig.json': tsconfigSource({ compilerOptions: { rootDir: 'api' } }) })
-    const result = await ctx.createLayoutThrow()
-    expect(result.sourceRoot).toMatchInlineSnapshot(`"__DYNAMIC__/api"`)
+  it('uses the value in tsconfig compilerOptions.rootDir if present', async () => {
+    ctx.setup({ 'tsconfig.json': tsconfigSource({ compilerOptions: { rootDir: 'api' } }), 'api/app.ts': '' })
+    const res = await ctx.createLayout().then(rightOrThrow)
+    expect(res.sourceRoot).toMatchInlineSnapshot(`"__DYNAMIC__/api"`)
   })
 })
 
-it('fails if empty file tree', async () => {
-  ctx.setup()
-
-  try {
-    await ctx.createLayoutThrow()
-  } catch (err) {
-    expect(err.message).toContain("Path you want to find stuff in doesn't exist")
-  }
-})
-
 describe('tsconfig', () => {
   beforeEach(() => {
     ctx.setup({ 'app.ts': '' })
   })
 
+  it('fails if tsconfig settings does not lead to matching any source files', async () => {
+    ctx.fs.remove('app.ts')
+    const res = await ctx.createLayout().then(leftOrThrow)
```

Comment:
Hit a type/code design problem tonight.

The type here is `res`. What we want (in general, maybe here specifically we don't care much) is a union of the possible errors. Ideally, every new `left` return in the nested codebase would "bubble up" and automatically appear here.

However, for that to happen we need to rely on return-type inference for all functions in the hierarchy. This is something I've in the past been semi-against because up-front func contracts  help with readability (otherwise constantly hovering to see return type) and also "trap" violations closer to source.

The latter point above is the one I'm most fearful of violating. If we have say 10 functions in  a hierarchy, and one returns the wrong thing, we'll have to go hunting through up to 10 functions to find out where the bad implementation is.

Maybe in practice this won't be such a problem! I'm not sure.

Would would be ideal is if we could do this:

```ts
function someFunc(...): Promise<Either<infer, Result>>
``` 

It seems that would give us the best of both worlds ^

Alas, no... Might open an issue on TS to propose it (though don't expect it to seriously work out).